storage/sql/coresql/testonly/checks.go

```diff
@@ -0,0 +1,114 @@
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+)
+
+// CheckLeafContents compares two log leaves to see if they match. It fails a test with
+// a descriptive message if not.
+func CheckLeafContents(leaf *trillian.LogLeaf, seq int64, rawHash, hash, data, extraData []byte, t *testing.T) {
+	if got, want := leaf.MerkleLeafHash, hash; !bytes.Equal(got, want) {
+		t.Fatalf("Wrong leaf hash in returned leaf got\n%v\nwant:\n%v", got, want)
+	}
+
+	if got, want := leaf.LeafIdentityHash, rawHash; !bytes.Equal(got, want) {
+		t.Fatalf("Wrong raw leaf hash in returned leaf got\n%v\nwant:\n%v", got, want)
+	}
+
+	if got, want := seq, leaf.LeafIndex; got != want {
+		t.Fatalf("Bad sequence number in returned leaf got: %d, want:%d", got, want)
+	}
+
+	if got, want := leaf.LeafValue, data; !bytes.Equal(got, want) {
+		t.Fatalf("Unxpected data in returned leaf. got:\n%v\nwant:\n%v", got, want)
+	}
+
+	if got, want := leaf.ExtraData, extraData; !bytes.Equal(got, want) {
+		t.Fatalf("Unxpected data in returned leaf. got:\n%v\nwant:\n%v", got, want)
+	}
+}
+
+// LeafInBatch tests if a log leaf is contained in a slice of log leaves.
+func LeafInBatch(leaf *trillian.LogLeaf, batch []*trillian.LogLeaf) bool {
+	for _, bl := range batch {
+		if bytes.Equal(bl.LeafIdentityHash, leaf.LeafIdentityHash) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// EnsureAllLeavesDistinct checks that the supplied leaves do not contain any duplicate
+// LeafIdentityHash Values.	All the leaf hashes should be distinct because the leaves were
+// created with distinct leaf data. If only we had maps with slices as keys or sets or pretty
+// much any kind of usable data structures we could do this properly.
+func EnsureAllLeavesDistinct(leaves []*trillian.LogLeaf, t *testing.T) {
+	for i := range leaves {
+		for j := range leaves {
+			if i != j && bytes.Equal(leaves[i].LeafIdentityHash, leaves[j].LeafIdentityHash) {
+				t.Fatalf("Unexpectedly got a duplicate leaf hash: %v %v",
+					leaves[i].LeafIdentityHash, leaves[j].LeafIdentityHash)
+			}
+		}
+	}
+}
+
+// NodesAreEqual tests if two slices of storage nodes contain the same set of nodes.
+func NodesAreEqual(lhs []storage.Node, rhs []storage.Node) error {
+	if ls, rs := len(lhs), len(rhs); ls != rs {
+		return fmt.Errorf("different number of nodes, %d vs %d", ls, rs)
+	}
+	for i := range lhs {
+		if l, r := lhs[i].NodeID.String(), rhs[i].NodeID.String(); l != r {
+			return fmt.Errorf("NodeIDs are not the same,\nlhs = %v,\nrhs = %v", l, r)
+		}
+		if l, r := lhs[i].Hash, rhs[i].Hash; !bytes.Equal(l, r) {
+			return fmt.Errorf("Hashes are not the same for %s,\nlhs = %v,\nrhs = %v", lhs[i].NodeID.CoordString(), l, r)
+		}
+	}
+	return nil
+}
+
+// DiffNodes compares two slices of storage nodes and returns slices containing extra and
+// missing nodes (if any). If the slices contain the same data both returned slices will
+// be empty.
+func DiffNodes(got, want []storage.Node) ([]storage.Node, []storage.Node) {
```

Comment:
Yes they are expected to be distinct. I've made it fail if either slice contains duplicates.